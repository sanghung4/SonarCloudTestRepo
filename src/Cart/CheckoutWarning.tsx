import { Box, useScreenSize } from '@dialexa/reece-component-library';
import { useTranslation } from 'react-i18next';
import { WarningIcon } from 'icons';
import {
  CheckoutWarningContainer,
  CheckoutWarningTypography,
  CheckoutWarningMobileTypography
} from 'Cart/util/styled';

export default function Warning() {
  /**
   * Custom Hooks
   */
  const { t } = useTranslation();
  const { isSmallScreen } = useScreenSize();

  /**
   * Render
   */
  return (
    <CheckoutWarningContainer data-testid="checkout-warning-message">
      <Box component={WarningIcon} ml={1} color="secondary.main" />
      <Box flex={1} ml={1}>
        {isSmallScreen ? (
          <CheckoutWarningMobileTypography>
            {t('cart.checkoutWarning')}
          </CheckoutWarningMobileTypography>
        ) : (
          <CheckoutWarningTypography>
            {t('cart.checkoutWarning')}
          </CheckoutWarningTypography>
        )}
      </Box>
    </CheckoutWarningContainer>
  );
}
