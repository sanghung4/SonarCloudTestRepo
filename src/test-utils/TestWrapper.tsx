import React, { JSXElementConstructor, ReactElement, ReactNode } from 'react';

import { ThemeProvider } from '@dialexa/reece-component-library';
import { MockedProvider, MockedResponse } from '@apollo/client/testing';
import { OktaAuth } from '@okta/okta-auth-js';
import { render as jestRender, RenderResult } from '@testing-library/react';
import { I18nextProvider } from 'react-i18next';
import { MemoryRouter } from 'react-router-dom';

import { UserContextType, AuthContext } from 'AuthProvider';
import { Permission } from 'common/PermissionRequired';
import i18n from 'test-utils/mockI18N';
import {
  SelectedAccountsContext,
  SelectedAccountsContextType
} from 'providers/SelectedAccountsProvider';
import { CartContext, CartContextType } from 'providers/CartProvider';
import {
  CheckoutContext,
  CheckoutContextType,
  initialCheckoutContextValues
} from 'Checkout/CheckoutProvider';

/**
 * NOTE:
 * To setup mocks follow the examples here:
 * https://www.apollographql.com/docs/react/development-testing/testing/
 *
 * To get a successful response in your test (not just the loading state)
 * you can do the following:
 *
 * it('Should render', async () => {
 *   const { container } = renderMockRouter(<MyComponent />, mocks);
 *
 *   await waitFor(() => new Promise((res) => setTimeout(res, 0)));
 *
 *   expect(container).toMatchSnapshot();
 * });
 *
 */

/******************************/
/* Types                      */
/******************************/
type Props = {
  children?: ReactNode;
  authConfig?: Partial<UserContextType>;
  selectedAccountsConfig?: Partial<SelectedAccountsContextType>;
  cartConfig?: Partial<CartContextType>;
  checkoutConfig?: Partial<CheckoutContextType>;
  mocks?: ReadonlyArray<MockedResponse>;
};

type WrapperProps<TKey extends keyof Props> = {
  value: Props[TKey];
  children: React.ReactNode;
};

/******************************/
/* Defaults                   */
/******************************/
const defaultAuth: UserContextType = {
  authState: {
    isAuthenticated: true
  },
  isLoggingOut: false,
  oktaAuth: new OktaAuth({ issuer: 'http://test.com' }),
  profile: {
    userId: 'testuser',
    permissions: Object.values(Permission),
    isEmployee: false,
    isVerified: false
  }
};

const defaultSelectedAccounts: SelectedAccountsContextType = {
  clearAccounts: () => {},
  updateAccounts: () => {},
  updateShippingBranchId: () => {},
  loading: false,
  selectedAccounts: {
    billTo: { erpAccountId: '' },
    shipTo: { erpAccountId: '' },
    shippingBranchId: ''
  },
  isMincron: false,
  isEclipse: false
};

const defaultCartContextValue: CartContextType = {
  addItemToCart: () => {},
  addItemsToCart: () => {},
  addAllListItemsToCart: () => {},
  cart: undefined,
  cartLoading: false,
  checkingOutWithQuote: false,
  contractBranch: undefined,
  clearContract: () => {},
  clearQuote: () => {},
  deleteCartItems: () => new Promise(() => {}),
  disableAddToCart: false,
  deleteItem: () => {},
  getUserCart: () => new Promise(() => {}),
  itemCount: 0,
  itemLoading: undefined,
  lineNotes: {},
  releaseContractToCart: () => {},
  setContract: () => {},
  setLineNotes: () => {},
  setQuoteId: () => {},
  setQuoteData: () => {},
  setQuoteShipToId: () => {},
  setPreviousCart: () => {},
  setSelectedBranch: () => {},
  updateCart: undefined,
  updateDelivery: undefined,
  updateItemQuantity: () => {},
  updateWillCall: undefined,
  updateWillCallBranch: undefined,
  refreshCart: undefined,
  isWillCall: false,
  deleteCreditCardFromCart: () => new Promise(() => {})
};

/******************************/
/* Wrappers                   */
/******************************/
function SelectedAccountsProviderWrapper({
  value,
  children
}: WrapperProps<'selectedAccountsConfig'>) {
  if (value) {
    return (
      <SelectedAccountsContext.Provider
        value={{ ...defaultSelectedAccounts, ...value }}
      >
        {children}
      </SelectedAccountsContext.Provider>
    );
  }

  return <>{children}</>;
}

function CartProviderWrapper({ value, children }: WrapperProps<'cartConfig'>) {
  if (value) {
    return (
      <CartContext.Provider value={{ ...defaultCartContextValue, ...value }}>
        {children}
      </CartContext.Provider>
    );
  }

  return <>{children}</>;
}

function CheckoutProviderWrapper({
  value,
  children
}: WrapperProps<'checkoutConfig'>) {
  if (value) {
    return (
      <CheckoutContext.Provider
        value={{ ...initialCheckoutContextValues, ...value }}
      >
        {children}
      </CheckoutContext.Provider>
    );
  }

  return <>{children}</>;
}

function TestWrapper(props: Props) {
  return (
    <ThemeProvider>
      <I18nextProvider i18n={i18n}>
        <AuthContext.Provider value={{ ...defaultAuth, ...props.authConfig }}>
          <MockedProvider mocks={props.mocks ?? []}>
            <SelectedAccountsProviderWrapper
              value={props.selectedAccountsConfig}
            >
              <CartProviderWrapper value={props.cartConfig}>
                <CheckoutProviderWrapper value={props.checkoutConfig}>
                  {props.children}
                </CheckoutProviderWrapper>
              </CartProviderWrapper>
            </SelectedAccountsProviderWrapper>
          </MockedProvider>
        </AuthContext.Provider>
      </I18nextProvider>
    </ThemeProvider>
  );
}

function jestRenderFn(wrapped: JSX.Element) {
  return jestRender(wrapped, {
    wrapper: MemoryRouter as
      | JSXElementConstructor<{
          children: ReactElement<any, string | JSXElementConstructor<any>>;
        }>
      | undefined
  });
}

export function render(component: ReactNode, config?: Props) {
  const wrapped = <TestWrapper {...config}>{component}</TestWrapper>;
  return jestRenderFn(wrapped);
}
export type TestRender = RenderResult<
  typeof import('@testing-library/dom/types/queries'),
  Element | DocumentFragment
>;

export function renderWithComponent(component: ReactNode, config?: Props) {
  const wrapped = <TestWrapper {...config}>{component}</TestWrapper>;
  return {
    render: jestRenderFn(wrapped),
    component: wrapped
  };
}

export default TestWrapper;
