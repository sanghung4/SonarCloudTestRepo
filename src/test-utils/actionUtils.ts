import { fireEvent, findByTestId, act } from '@testing-library/react';
import userEvent from '@testing-library/user-event';

export const fillFormTextInput = async (testId: string, value: string) => {
  await act(async () => {
    const textInput = await findByTestId(document.body, `${testId}-input`);
    await fireEvent.change(textInput, { target: { value } });
  });
};

export const fillMaskedFormTextInput = async (
  testId: string,
  value: string
) => {
  await act(async () => {
    const textInput = await findByTestId(document.body, `${testId}-input`);
    await userEvent.type(textInput, value);
  });
};

export const clickButton = async (testId: string) => {
  await act(async () => {
    const button = await findByTestId(document.body, testId);
    await fireEvent.click(button);
  });
};

export const clickRadio = async (testId: string, optionIndex: number) => {
  await act(async () => {
    const radio = await findByTestId(
      document.body,
      `${testId}-radio-${optionIndex}`
    );
    await userEvent.click(radio);
  });
};

export const fillFormSelectInput = async (testId: string, value: string) => {
  await act(async () => {
    const selectInput = await findByTestId(document.body, `${testId}-input`);
    await fireEvent.change(selectInput, { target: { value } });
    await fireEvent.keyDown(window, { key: 'Enter' });
  });
};

export const focusInput = async (testId: string) => {
  await act(async () => {
    const input = await findByTestId(document.body, testId);
    await fireEvent.focus(input);
  });
};

export const changeInput = async (testId: string, value: string) => {
  await act(async () => {
    const textInput = await findByTestId(document.body, testId);
    await fireEvent.change(textInput, { target: { value } });
  });
};

export const blurInput = async (testId: string) => {
  await act(async () => {
    const input = await findByTestId(document.body, testId);
    await fireEvent.blur(input);
  });
};

export const keyboardDown = async (
  key: string,
  element?: Element | Node | Document | Window
) => {
  await act(async () => {
    await fireEvent.keyDown(element ?? window, { key });
  });
};
