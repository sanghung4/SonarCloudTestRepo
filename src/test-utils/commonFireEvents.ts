import { act, fireEvent } from '@testing-library/react';

export function actClickOnElement(element: HTMLElement) {
  act(() => {
    fireEvent.click(element);
  });
}
export function actUpdateInput(input: HTMLElement, value: string) {
  act(() => updateInput(input, value));
}
export function updateInput(input: HTMLElement, value: string) {
  fireEvent.focus(input);
  fireEvent.change(input, { target: { value } });
  fireEvent.blur(input);
}
