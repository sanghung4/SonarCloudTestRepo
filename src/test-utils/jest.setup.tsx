import React, { JSXElementConstructor, ReactElement, ReactNode } from "react";
import { MockedProvider, MockedResponse } from "@apollo/client/testing";
import { MemoryRouter } from "react-router-dom";
import { render as jestRender } from "@testing-library/react";

export const mockApollo = jest.mock(
  "@apollo/client/testing",
  () => MockedProvider
);

interface Props {
  children?: ReactNode;
  mocks?: ReadonlyArray<MockedResponse>;
}

function TestWrapper(props: Props) {
  return (
    <MockedProvider mocks={props.mocks ?? []}>{props.children}</MockedProvider>
  );
}

export function render(component: ReactNode, config?: Props) {
  const wrapped = <TestWrapper {...config}>{component}</TestWrapper>;
  return jestRender(wrapped, {
    wrapper: MemoryRouter as
      | JSXElementConstructor<{
          children: ReactElement<any, string | JSXElementConstructor<any>>;
        }>
      | undefined,
  });
}
