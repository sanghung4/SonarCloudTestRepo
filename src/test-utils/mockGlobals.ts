export function mockGoogle() {
  global.window.google = {
    maps: {
      // @ts-ignore
      places: {
        Autocomplete: jest.fn().mockImplementation(() => ({
          setFields: jest.fn(),
          setComponentRestrictions: jest.fn(),
          addListener: jest.fn()
        }))
      },
      Size: jest.fn(),
      Point: jest.fn()
    }
  };
}

export function mockGeolocation(lat: number, lng: number, err?: boolean) {
  // @ts-ignore
  global.navigator.geolocation = {
    getCurrentPosition: jest.fn().mockImplementation((onSuccess, onError) => {
      onSuccess({
        coords: {
          latitude: lat ?? 0,
          longitude: lng ?? 0
        }
      });
      if (err) {
        onError(new Error());
      }
    }),
    watchPosition: jest.fn().mockImplementation((onSuccess, onError) => {
      onSuccess({
        coords: {
          latitude: lat ?? 0,
          longitude: lng ?? 0
        }
      });
      if (err) {
        onError(new Error());
      }
    })
  };
}

export function mockLocalStorage() {
  // @ts-ignore
  global.window.localStorage = class mockLocalStorage {
    store: { [key: string]: string };
    constructor() {
      this.store = {};
    }
    clear() {
      this.store = {};
    }
    get length() {
      return Object.keys(this.store).length;
    }
    getItem(key: string) {
      return this.store[key] || null;
    }
    key(i: number) {
      return Object.keys(this.store)[i] || null;
    }
    removeItem(key: string) {
      delete this.store[key];
    }
    setItem(key: string, value: string) {
      this.store[key] = String(value);
    }
  };
}
