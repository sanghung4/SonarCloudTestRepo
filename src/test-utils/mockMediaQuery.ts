import mediaQuery from 'css-mediaquery';

function createMatchMedia(width: number) {
  return (query: string): MediaQueryList => ({
    matches: mediaQuery.match(query, { width }),
    media: query,
    onchange: null,
    addListener: () => jest.fn(),
    removeListener: () => jest.fn(),
    addEventListener: jest.fn(),
    removeEventListener: jest.fn(),
    dispatchEvent: jest.fn()
  });
}

export const breakpoints = {
  mobile: 414,
  desktop: 1200
};

export const setBreakpoint = (breakpoint: 'mobile' | 'desktop' = 'desktop') => {
  Object.defineProperty(window, 'innerWidth', {
    writable: true,
    configurable: true,
    value: breakpoints[breakpoint]
  });
  window.dispatchEvent(new Event('resize'));

  window.matchMedia = createMatchMedia(window.innerWidth);
};
