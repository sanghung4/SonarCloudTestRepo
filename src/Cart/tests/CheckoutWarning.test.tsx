import CheckoutWarning from 'Cart/CheckoutWarning';
import { setBreakpoint } from 'test-utils/mockMediaQuery';
import { render } from 'test-utils/TestWrapper';

describe('Cart - Warning', () => {
  it('expect to match snapshop on desktop', () => {
    setBreakpoint('desktop');
    const { container } = render(<CheckoutWarning />);
    expect(container).toMatchSnapshot();
  });

  it('expect to match snapshop on mobile', () => {
    setBreakpoint('mobile');
    const { container } = render(<CheckoutWarning />);
    expect(container).toMatchSnapshot();
  });
});
