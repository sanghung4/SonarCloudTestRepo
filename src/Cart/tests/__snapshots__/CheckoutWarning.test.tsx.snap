// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Cart - Warning expect to match snapshop on desktop 1`] = `
<div>
  <div
    class="MuiBox-root css-jhm0j8"
    data-testid="checkout-warning-message"
  >
    <svg
      class="MuiBox-root css-1k8admq"
      sx="[object Object]"
    >
      warning.svg
    </svg>
    <div
      class="MuiBox-root css-on40yh"
    >
      <p
        class="MuiTypography-root MuiTypography-body1 css-40zv4g-MuiTypography-root"
      >
        You have more than 200 items in your cart! It may take a minute to finish checking out.
      </p>
    </div>
  </div>
</div>
`;

exports[`Cart - Warning expect to match snapshop on mobile 1`] = `
<div>
  <div
    class="MuiBox-root css-jhm0j8"
    data-testid="checkout-warning-message"
  >
    <svg
      class="MuiBox-root css-1k8admq"
      sx="[object Object]"
    >
      warning.svg
    </svg>
    <div
      class="MuiBox-root css-on40yh"
    >
      <p
        class="MuiTypography-root MuiTypography-body1 css-pu2qy7-MuiTypography-root"
      >
        You have more than 200 items in your cart! It may take a minute to finish checking out.
      </p>
    </div>
  </div>
</div>
`;
